### PR TITLE
adapter: Add user permissions framework

### LINKED ISSUE #15520: Add a mz_introspection system user
### Feature request

We should add a new system user named `mz_introspection`. This user will be allowed limited read only access to the system via the internal SQL port.

Here is a proposed set of restrictions:

  * The user can only log in via the internal SQL port.
  * The user can execute only `SELECT` or `SUBSCRIBE` statements that involve objects in the `mz_internal`, `mz_catalog`, or `pg_catalog` schemas.
  * The user cannot execute DDL statements or DML statements, besides those allowed in the last bullet point.
  * The user can only execute queries on the system clusters (`mz_system` or `mz_introspection`, but possibly others in the future).

There will be no authentication required, beyond having access to the internal SQL port. I.e., you'll be able to freely connect with `psql -U mz_system -p 6877` or `psql -U mz_introspection -p 6877`. The idea is to reduce mistakes due to typos, etc, not to protect against malicious actors.

cc @jonathan-k-shapiro @chaas

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds a framework for specifying and enforcing user level
permissions. The permissions indicate what clusters and schemas a user
is allowed to interact with as well as what actions that user is
allowed to perform. The permissions check is done right after statement
planning, but before sequencing.

The mz_introspection user is given a set of permissions to restrict
what it's allowed to do.

The permissions framework is non-internal users to prevent users from
interacting with the mz_system cluster, which was previously done in an
incomplete way.

Closes #15520

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by jkosh44** [comment]
I'm planning on adding a bunch of unit tests to the `permissions.rs` file. However, I'm not really completely sold that this is the right approach and there is some rough edges IMO, so I thought it'd be better if people took a look.

**Review by mjibson — COMMENTED** [review]
The meaning of Blocklist is sufficiently non-obvious it should maybe change. For the `Default` user permission, on the `mz_system` compute instance, the user is blocked from doing the empty set of actions. Does that mean the user can do anything or nothing? Are schemas the same wrt the empty set? Negation gets real weird.

We might consider a solution where users start with no permissions and can only be granted permissions to do things.

I think that if we want to solve the linked issue today, we should perhaps hard code some permissions checking until we have a more robust system in place, because we definitely know the hard coded stuff is bad (as noted by one of the removed comments in this diff).

**Comment by jkosh44** [comment]
@mjibson Going through your comments now. A couple things that I wasn't ecstatic about when submitting this are below. If you have any strong opinions then let me know, otherwise feel free to ignore.
- I'm not a huge fan of the schema list and cluster list being strings and not Ids. However, Ids make it difficult to hard code a permission for specific users since the Ids aren't known at compile time. We could take a similar approach that `BuiltinType`s take which is to have a generic parameter indicate whether it uses a name or id (`BuiltinType<NameReference>` vs `BuiltinType<IdReference>`).
- The functions in `permissions.rs`: `extract_target_schema`, `extract_target_compute_instance`, and `uses_compute_instance` all seem a bit flaky.

File: src/adapter/src/catalog/builtin.rs [hunk 1]
```diff
@@ -2465,16 +2465,16 @@ ON mz_catalog.mz_secrets (schema_id)",
 };
 
 pub static MZ_SYSTEM_ROLE: Lazy<BuiltinRole> = Lazy::new(|| BuiltinRole {
-    name: &*SYSTEM_USER.name,
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/adapter/src/catalog/builtin.rs:2468]: I switched them.

**Comment by jkosh44** [comment]
I'm closing this PR in favor of implementing something in the future that is closer to Postgres's Privilege framework (https://www.postgresql.org/docs/current/ddl-priv.html).

**Comment by benesch** [comment]
> I'm closing this PR in favor of implementing something in the future that is closer to Postgres's Privilege framework (https://www.postgresql.org/docs/current/ddl-priv.html).

Taking a small step towards generalized RBAC sounds great to me!

**Pull request closed without merging**